# Tehtävä 6: Optimaalinen ostoerä

# Osa 1

Lähtötiedot

Yritys ostaa vuosittain 2 000 termostaattia hintaan 20 000 euroa/kappale. Tilauskustannukset ovat 5 000 euroa/tilaus ja varastointikustannukset 15 % varaston ostohintaisesta arvosta.

Yritys ostaa tavaraa optimaalisen ostoerän mukaan.

Minkä suuruinen on optimaalinen ostoerä? (50 %)
Myyjä tarjoaa 3 %:n alennusta ostohintaan, jos tilausmääräksi muutetaan 2 000 kappaletta eli koko vuoden tarve ostetaan kerralla. Kannattaako tarjous hyväksyä? (50 %)

In [13]:
import math
from scipy.stats import norm



def laske_eoq(demand, order_cost, holding_cost_rate, unit_price):
    holding_cost = holding_cost_rate * unit_price
    eoq = math.sqrt((2 * demand * order_cost) / holding_cost)
    return eoq

def laske_kokonaiskustannus(demand, order_quantity, order_cost, holding_cost_rate, unit_price, discount=0):
    effective_unit_price = unit_price * (1 - discount)
    holding_cost = holding_cost_rate * effective_unit_price
    order_frequency = demand / order_quantity
    total_ordering_cost = order_frequency * order_cost
    total_holding_cost = (order_quantity / 2) * holding_cost
    total_purchase_cost = demand * effective_unit_price
    total_cost = total_ordering_cost + total_holding_cost + total_purchase_cost
    return total_cost

# Lähtötiedot
annual_demand = 2000
unit_price = 20000
order_cost = 5000
holding_cost_rate = 0.15

eoq = laske_eoq(annual_demand, order_cost, holding_cost_rate, unit_price)

discount_order_quantity = 2000
discount_rate = 0.03
eoq_cost = laske_kokonaiskustannus(annual_demand, eoq, order_cost, holding_cost_rate, unit_price)
discount_cost = laske_kokonaiskustannus(annual_demand, discount_order_quantity, order_cost, holding_cost_rate, unit_price, discount=discount_rate)



def laske_varmuusvarasto(std_dev, lead_time, service_level):
    z_score = norm.ppf(service_level)
    safety_stock = z_score * std_dev * math.sqrt(lead_time)
    return safety_stock

def laske_tilauspiste(avg_demand, lead_time, safety_stock):
    reorder_point = (avg_demand * lead_time) + safety_stock
    return reorder_point

# Lähtötiedot

weekly_demand = 230
weekly_std_dev = 67
lead_time = 1  # viikko
service_level = 0.85

safety_stock = laske_varmuusvarasto(weekly_std_dev, lead_time, service_level)
reorder_point = laske_tilauspiste(weekly_demand, lead_time, safety_stock)

# Tulostus
print(f"Optimaalinen ostoerä (EOQ): {eoq:.2f} kpl")
print(f"EOQ-mallin kokonaiskustannus: {eoq_cost:.2f} euroa")
print(f"3 % alennuksen mukaisen tilauksen kokonaiskustannus: {discount_cost:.2f} euroa")
print(f"Kannattaako tarjous? {'Kyllä' if discount_cost < eoq_cost else 'Ei'}")
print(f"Suositeltu varmuusvarasto: {safety_stock:.2f} kpl")
print(f"Tilauspiste: {reorder_point:.2f} kpl")

Optimaalinen ostoerä (EOQ): 81.65 kpl
EOQ-mallin kokonaiskustannus: 40244948.97 euroa
3 % alennuksen mukaisen tilauksen kokonaiskustannus: 41715000.00 euroa
Kannattaako tarjous? Ei
Suositeltu varmuusvarasto: 69.44 kpl
Tilauspiste: 299.44 kpl


# Osa 1 Yhteenveto

Selvitettiin yrityksen optimaalinen ostoerä ja arvioitiin, kannattaako myyjän tarjoama 3 % alennus, jos koko vuoden tarve tilataan kerralla.

Yyrityksen optimaalinen ostoerä on 81.65 kappaletta, mikä minimoi varastointi- ja tilauskustannukset. 

Kun vertailtiin EOQ-mallin ja alennuksen mukaisia kokonaiskustannuksia, havaittiin, että EOQ-mallin kokonaiskustannus on 40 244 948.97 euroa, kun taas alennuksen mukaiset kokonaiskustannukset ovat 41 715 000.00 euroa. 

Koska alennuksen mukaiset kokonaiskustannukset ovat suuremmat kuin EOQ-mallin mukaiset kustannukset, tarjousta ei kannata hyväksyä.

Lisäksi kuinka suuri varasto tulisi pitää kysynnän vaihteluiden varalta. Laskelmien mukaan suositeltu varmuusvarasto on 69.44 kappaletta, ja tilaus kannattaa tehdä, kun varastossa on 299.44 kappaletta jäljellä. Tuotteita on saatavilla myös kysynnän vaihteluissa ja estää varaston loppumisen kriittisellä hetkellä.


# Osa 2

Lähtötiedot
Yritys myy keskimäärin 230 tietyn merkkistä ja tyyppistä termostaattia viikossa. 
    
Termostaatin viikkokysynnän keskihajonta on 67 kpl. 

Tilaukset perustuvat varaston jatkuvaan seurantaan. 
    
Toimitusaika on 1 viikko. 

Palvelutasoksi halutaan 85 %. 

Yhden navigaattorin pitokustannukseksi arvioidaan 2 euroa viikossa.

a) Kuinka suurta varmuusvarastoa suosittelet?
b) Kuinka suuren varastomäärän kohdalla uusi tilaus kannattaisi tehdä.


In [14]:
import math
from scipy.stats import norm

def laske_varmuusvarasto(std_dev, lead_time, service_level):
    z_score = norm.ppf(service_level)  # Z-arvo palvelutasolle
    safety_stock = z_score * std_dev * math.sqrt(lead_time)
    return safety_stock

def laske_tilauspiste(avg_demand, lead_time, safety_stock):
    reorder_point = (avg_demand * lead_time) + safety_stock
    return reorder_point

# Lähtötiedot
weekly_demand = 230        # Keskimääräinen viikkokysyntä
weekly_std_dev = 67        # Viikkokysynnän keskihajonta
lead_time = 1              # Toimitusaika viikkoina
service_level = 0.85       # Palvelutaso (85 %)

safety_stock = laske_varmuusvarasto(weekly_std_dev, lead_time, service_level)
reorder_point = laske_tilauspiste(weekly_demand, lead_time, safety_stock)

# Tulostus
print(f"Suositeltu varmuusvarasto: {safety_stock:.2f} kpl")
print(f"Tilauspiste: {reorder_point:.2f} kpl")


Suositeltu varmuusvarasto: 69.44 kpl
Tilauspiste: 299.44 kpl


# Osa 2 Yhteenveto 

Kuinka suuri varmuusvarasto yrityksen tulisi pitää ja millä varastotasolla uuden tilauksen tekeminen on suositeltavaa.  

Laskelmien perusteella suositeltu varmuusvarasto on 69.44 kappaletta. 

Tämä varmuusvarasto auttaa varautumaan kysynnän vaihteluihin ja estää varastojen loppumisen odottamattomien tilanteiden aikana.  

Lisäksi määritettiin tilauspiste, eli varastotaso, jonka kohdalla uusi tilaus tulisi tehdä. 

Tilaus kannattaa tehdä, kun varastossa on 299.44 kappaletta jäljellä. 

Tämä varmistaa, että tuotteita on saatavilla myös toimitusajan aikana ja ehkäisee mahdollisia myyntikatkoksia.  

Näin yritys voi optimoida varastonhallintansa, pienentää kustannuksia ja varmistaa, että tuotteita on saatavilla asiakkaille oikeaan aikaan.